common/src/main/java/org/jboss/pnc/common/Configuration.java

```diff
@@ -33,19 +37,25 @@
      * @return Loaded configuration
      * @throws ConfigurationParseException Thrown if configuration file couldn't be loaded or parsed
      */
-    public T getModuleConfig(Class<T> moduleClass) throws ConfigurationParseException {
+    @SuppressWarnings("unchecked")
+    public <T extends AbstractModuleConfig> T getModuleConfig(Class<T> moduleClass) throws ConfigurationParseException {
```

Comment:
@jbartece Slightly different - with a small synchronized block on `this`. Take a look at the example:
```
// Works with acquire/release semantics for volatile
// Broken under current semantics for volatile
  class Foo {
        private volatile Helper helper = null;
        public Helper getHelper() {
            if (helper == null) {
                synchronized(this) {
                    if (helper == null)
                        helper = new Helper();
                }
            }
            return helper;
        }
    }
```
taken from http://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

Now the point of this whole trick is very simple - if the field is already initialized - we want to return quickly without acquiring a monitor (to be specific monitor on `this`). Now if the field is not initialized, we obtain the monitor and check it once again - in that brief moment some other thread could initialize it and because we use synchronization - there is a "happens-before" relation between those 2 events. Finally, if the field is still not initialized - we do the initialization with holding a monitor.